### PR TITLE
QMJ partial impl

### PR DESCRIPTION
**Pull request by AlexIIL**
Misc work. I haven't finished yet, but we're a lot closer to quilt mods being able to load than before.
image
(*Notice the "quilt" icon next to the jar :)* )

I haven't even written *fallback* error descriptions yet, hence the todo's.

I'll merge this soon, although I'm not sure if we need to squash it?

### CODE REVIEW

**Comment by AlexIIL** [comment]
(This is only a PR rather than a direct push since this steps pretty heavily on QMJ as a whole).

File: src/experimental/java/org/quiltmc/loader/impl/VersionConstraintImpl.java
Comment (TheGlitch76) [inline src/experimental/java/org/quiltmc/loader/impl/VersionConstraintImpl.java:118]: This returns false for a.matches(a)

File: src/experimental/java/org/quiltmc/loader/api/ConvertibleModMetadata.java
Comment (TheGlitch76) [inline src/experimental/java/org/quiltmc/loader/api/ConvertibleModMetadata.java:8]: this *is* gross, but it's fine because abstracting all this away will be its own project

Comment (i509VCB) [reply src/experimental/java/org/quiltmc/loader/api/ConvertibleModMetadata.java:8]: This shouldn't be api imo.

File: src/experimental/java/org/quiltmc/loader/impl/VersionConstraintImpl.java
Comment (i509VCB) [inline src/experimental/java/org/quiltmc/loader/impl/VersionConstraintImpl.java:114]: This should be dev only really

File: src/experimental/java/org/quiltmc/loader/api/ConvertibleModMetadata.java
Comment (AlexIIL) [reply src/experimental/java/org/quiltmc/loader/api/ConvertibleModMetadata.java:8]: Actually it would probably be cleaner if this was moved outside of api too :p

File: src/experimental/java/org/quiltmc/loader/impl/VersionConstraintImpl.java
Comment (AlexIIL) [reply src/experimental/java/org/quiltmc/loader/impl/VersionConstraintImpl.java:114]: Yes, that's a good point.

File: src/experimental/java/org/quiltmc/loader/impl/VersionConstraintImpl.java
Comment (AlexIIL) [reply src/experimental/java/org/quiltmc/loader/impl/VersionConstraintImpl.java:120]: How?

`VersionConstraintImpl` doesn't implement `Version`, so it's not possible to pass the constaint back into itself (`VersionConstraintImpl.matches(Version)`).

(Although I do think it's a bit off that we can't handle string versions here, however that's what the QMJ spec suggests?)

**Review by TheGlitch76 — APPROVED** [review]
Good enough for me. This is all experimental API, let's get a prototype in.

**Comment by AlexIIL** [comment]
Should we wait for a second review? (If we started following https://github.com/QuiltMC/rfcs/pull/25 now at least, we'd need a second one).

Either way I'm not going to make any further changes for now.

**Comment by TheGlitch76** [comment]
i5 gave me permission to push whatever into the qmj branch so I imagine it's fine. Might be worth pinging him on discord?

**Comment by AlexIIL** [comment]
From @i509VCB in #loader-general:
> [21:11] i509VCB: I'd say it's fine
[21:11] i509VCB: I am going to do a big testing operation before we unveil it for use

So I'll merge this now.

**Pull request merged**
